open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/App2AppAuthenticator.java

```diff
@@ -0,0 +1,214 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean canHandle(HttpServletRequest httpServletRequest) {
+
+        /*
+        App2App authenticates the user in one step depending on the app_auth_key,
+        Hence it's mandatory to have the required parameter app_auth_key.
+         */
+        return !StringUtils.isBlank(httpServletRequest.getParameter(
+                App2AppAuthenticatorConstants.DEVICE_VERIFICATION_TOKEN_IDENTIFIER));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getContextIdentifier(HttpServletRequest request) {
+
+        return request.getParameter(App2AppAuthenticatorConstants.SESSION_DATA_KEY);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request,
+                                                 HttpServletResponse response,
+                                                 AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        /*
+            App2App authenticator does not support initiating authentication request,
+            Hence authentication process will be terminated.
+         */
+        log.error(App2AppAuthenticatorConstants.INITIALIZATION_ERROR_MESSAGE);
+        throw new AuthenticationFailedException(
+                App2AppAuthenticatorConstants.DEVICE_VERIFICATION_TOKEN_MISSING_ERROR_MESSAGE);
+    }
+
+    /**
+     * Retrieves the public key associated with a device and user.
+     *
+     * @param deviceID    The identifier of the device for which the public key is requested.
+     * @param authenticatedUser  the authenticated user for this request
+     * @return            The public key associated with the specified device and user.
+     * @throws UserStoreException                If an error occurs while accessing user store.
+     * @throws PushDeviceHandlerServerException  If an error occurs on the server side of the push device handler.
+     * @throws PushDeviceHandlerClientException  If an error occurs on the client side of the push device handler.
+     */
+    private String getPublicKeyByDeviceID(String deviceID, AuthenticatedUser authenticatedUser)
+            throws UserStoreException, PushDeviceHandlerServerException, PushDeviceHandlerClientException,
+            OpenBankingException {
+
+        DeviceHandler deviceHandler = getDeviceHandler();
+        UserRealm userRealm = App2AppAuthUtils.getUserRealm(authenticatedUser);
+        String userID = App2AppAuthUtils.getUserIdFromUsername(authenticatedUser.getUserName(), userRealm);
+        return App2AppAuthUtils.getPublicKey(deviceID, userID, deviceHandler);
+    }
+
+    /**
+     * Method to get a DeviceHandler implementation.
+     *
+     * @return an implementation of DeviceHandler
+     */
+    private DeviceHandler getDeviceHandler() {
+
+        if (deviceHandler == null) {
+            deviceHandler = new DeviceHandlerImpl();
```

Comment:
Shall we consider moving this line to the constructor? whenever a new App2AppAuthenticator is created, the deviceHandler ends up being null, which would prevent this from working as a singleton.

Reply:
Resolved in [cbf2bbd](https://github.com/wso2/financial-open-banking/pull/67/commits/cbf2bbd7f62fb95d3d0e721f80e52b914d8fe80d).